languages/ruby/exercises/concept/enumeration/.docs/introduction.md

```diff
@@ -0,0 +1,40 @@
+Enumeration is the act of stepping through a collection (Array, Hash, etc) and performing some action on each object.
+Enumeration is a key concept in Ruby and is used for sorting (`sort_by`), grouping (`group_by`), mapping (`map`), reducing (`reduce`), and much more. You'll most frequently see enumeration as the idiomatic way iterating through collections rather than using loops.
+
+A simple enumeration to print each word in an array would look like this:
+
+```ruby
+words = %w[the cat sat on the mat]
+words.each { |word| print word }
+
+#=> the
+#=> cat
+#=> ...
+#=> mat
+```
+
+In this example, we have called the Array's `each` method and passed in a _block_, which takes one parameter (`word`) and uppercases it. Blocks are small groupings of statements that can be executed multiple times. They can be thought of as closures or anonymous functions. Blocks are defined using the `do...end` syntax (above), or the `{}` (below). The styles are interchanagable and differening opinions exist about when each should be used.
+
+We can also chain enumarable methods. For example, we can chain `.with_index` onto `each` to print out the index of an object as well as it's value:
+
+```ruby
+words = %w[the cat sat on the mat]
+words.each.with_index { |word, index| print "#{index}: #{word}" }
+
+#=> 0. the
+#=> 1. cat
+#=> ...
+#=> 5. mat
+```
+
+Enumerating hashes is exactly the same as enumerating arrays, except that the block receives two parameters: the key and the value:
+
+```
+pet_names = {cat: "bob", horse: "caris", mouse: "arya"}
+words.each { |animal, name| ... }
+
+# The two params should be put in brakets when chaining
+words.each.with_index { |(animal, name), index| ... }
+```
+
+The various enumeration methods can be see in the docs for the the [Enumerable Module](https://ruby-doc.org/core-2.7.1/Enumerable.html) which is part of Arrays, Hashes, etc.
```

Comment:
I think the intention with that quote is that we're not pushing people to have to read other blogs to understand the concept. I think linking to the official docs, which contain the methods that someone needs to discover to solve the exercise, is a "friendly" action rather than "harmful" one. I **believe** I've been consistent in that opinion, but happy to be told where I'm being a total hypocrite :)

Reply:
> I think linking to the official docs, which contain the methods that someone needs to discover to solve the exercise, is a "friendly" action rather than "harmful" one.

Well, in the case of Ruby documentation, I feel like I could make the case that it is harmful :P My main "worry" here is that we link to the entire document, not specific methods (which we _can_ do in the `hints.md` file). There is just so much content there that I feel it could easily feel overwhelming.

Maybe I remember this wrong, but I thought we wanted to mention the official names but not link to them here, but only in the `hints.md` file. This is also how I've reviewed all the existing tracks' PR's, so I we change this we should probably be making this more clear in the documentation.

Reply:
But maybe I'm overthinking this and linking to documentation is fine.